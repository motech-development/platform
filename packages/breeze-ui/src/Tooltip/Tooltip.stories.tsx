import { withA11y } from '@storybook/addon-a11y';
import { select, text, withKnobs } from '@storybook/addon-knobs';
import { storiesOf } from '@storybook/react';
import React from 'react';
import BaseStyles from '../BaseStyles/BaseStyles';
import Button from '../Button/Button';
import Tooltip from './Tooltip';

const stories = storiesOf('Tooltip', module);
const colour = {
  Danger: 'danger',
  Primary: 'primary',
};
const placement = {
  Bottom: 'bottom',
  Left: 'left',
  Right: 'right',
  Top: 'top',
};

stories.addDecorator(withA11y);
stories.addDecorator(withKnobs);

stories.add('Basic tooltip', () => (
  <>
    <BaseStyles />

    <Tooltip
      id="test"
      parent={
        <Button colour="primary" size="lg">
          Hover over me
        </Button>
      }
      placement={
        select('Placement', placement, 'bottom') as
          | 'bottom'
          | 'left'
          | 'right'
          | 'top'
      }
      colour={select('Colour', colour, 'primary')}
      message={text('Text', 'This is a tooltip!')}
    />
  </>
));
