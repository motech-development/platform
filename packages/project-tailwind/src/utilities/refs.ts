import { LegacyRef, MutableRefObject, useMemo, useRef } from 'react';

/**
 * Check if ref is a mutable ref object
 *
 * @param ref - Reference object or function
 *
 * @returns If ref is a mutable ref object
 */
function isMutableRefObject<T>(
  ref: MutableRefObject<T> | LegacyRef<T>,
): ref is MutableRefObject<T | null> {
  return ref != null;
}

/** useMergeRefs arguments */
type TUseMergeRefsInput<T> = Array<MutableRefObject<T> | LegacyRef<T>>;

/**
 * Combines multiple refs into a single memoized callback instance
 *
 * @param refs - Reference object or function
 *
 * @returns Combined ref object
 */
export default function useMergeRefs<T>(...refs: TUseMergeRefsInput<T>) {
  const targetRef = useRef<T>(null);

  return useMemo(() => {
    if (refs.every((ref) => ref == null)) {
      return null;
    }

    refs.forEach((ref) => {
      if (typeof ref === 'function') {
        ref(targetRef.current);
      } else if (isMutableRefObject(ref)) {
        // eslint-disable-next-line no-param-reassign
        ref.current = targetRef.current;
      }
    });

    return targetRef;
  }, [refs]);
}
