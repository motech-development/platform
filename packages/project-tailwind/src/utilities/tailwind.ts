import clsx, { ClassValue } from 'clsx';

/** Alignment options */
export enum Alignment {
  /** Centre */
  CENTRE = 'centre',
  /** Left */
  LEFT = 'left',
  /** Right */
  RIGHT = 'right',
}

/** Alignment type */
export type TAlignment = `${Alignment}`;

/** Sizing options */
export enum Sizing {
  /** Large */
  LG = 'lg',
  /** Medium */
  MD = 'md',
  /** None */
  NONE = 'none',
  /** Small */
  SM = 'sm',
}

/** Sizing type */
export type TSizing = `${Sizing}`;

/** Theming options */
export enum Themes {
  /** Danger */
  DANGER = 'danger',
  /** Primary */
  PRIMARY = 'primary',
  /** Secondary */
  SECONDARY = 'secondary',
  /** Success */
  SUCCESS = 'success',
  /** Warning */
  WARNING = 'warning',
}

/** Theme type */
export type TTheme = `${Themes}`;

/** Class names options */
type TCreateStylesClassNames = {
  /** Universal classes to apply */
  classNames?: ClassValue[];
  /** Sizing specific classes */
  sizing?: {
    [name in Sizing]?: ClassValue[];
  };
  /** Theme specific classes */
  theme?: {
    [name in TTheme]?: ClassValue[];
  };
};

/**
 * Tailwind hook
 *
 * @param theme - The theme to apply
 * @param sizing - Sizing to apply
 *
 * @returns Tailwind helper utilities
 */
export function useTailwind(theme: TTheme, sizing: TSizing) {
  /**
   * Creates css styles to apply
   *
   * @param classNames - Classes to be applied
   *
   * @returns Class names
   */
  function createStyles(classNames: TCreateStylesClassNames) {
    let finalStyles: ClassValue[] = [];

    if (classNames.classNames) {
      finalStyles = [...classNames.classNames];
    }

    if (sizing && classNames.sizing) {
      const styles = classNames.sizing[sizing];

      if (!styles) {
        throw new Error('No classes defined for this sizing.');
      }

      finalStyles = [...finalStyles, ...styles];
    }

    if (theme && classNames.theme) {
      const styles = classNames.theme[theme];

      if (!styles) {
        throw new Error('No classes defined for this theme.');
      }

      finalStyles = [...finalStyles, ...styles];
    }

    return clsx(finalStyles);
  }

  /**
   * Creates text alignment classes
   *
   * @param alignment - Text alignment
   *
   * @returns Class names
   */
  function createTextAlignmentStyles(alignment: TAlignment) {
    switch (alignment) {
      case Alignment.CENTRE:
        return 'text-center';
      case Alignment.RIGHT:
        return 'text-right';
      case Alignment.LEFT:
      default:
        return 'text-left';
    }
  }

  return {
    createStyles,
    createTextAlignmentStyles,
  };
}
