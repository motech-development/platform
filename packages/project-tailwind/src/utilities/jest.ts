/* eslint-disable import/no-extraneous-dependencies */
import { render, RenderResult } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { UserEvent } from '@testing-library/user-event/dist/types/setup/setup';
import { ReactElement } from 'react';
import { Sizing, Themes } from './tailwind';

export const themes = [
  {
    theme: Themes.DANGER,
  },
  {
    theme: Themes.PRIMARY,
  },
  {
    theme: Themes.SECONDARY,
  },
  {
    theme: Themes.SUCCESS,
  },
  {
    theme: Themes.WARNING,
  },
];

export const sizing = [
  {
    size: Sizing.LG,
  },
  {
    size: Sizing.MD,
  },
  {
    size: Sizing.NONE,
  },
  {
    size: Sizing.SM,
  },
];

type TSetup = RenderResult & {
  user: UserEvent;
};

/**
 * Combines Testing Library render function with user events setup
 *
 * @param ui - React element to test
 *
 * @returns Rendered component and test user
 */
export function setup(ui: ReactElement): TSetup {
  return {
    user: userEvent.setup({
      delay: null,
    }),
    ...render(ui),
  };
}
