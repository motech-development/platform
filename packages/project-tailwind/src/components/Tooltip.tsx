import {
  arrow,
  autoUpdate,
  flip,
  MiddlewareData,
  offset,
  Placement,
  shift,
  Side,
  useDismiss,
  useFloating,
  useFocus,
  useHover,
  useInteractions,
  useRole,
} from '@floating-ui/react';
import {
  ComponentPropsWithoutRef,
  ReactNode,
  useEffect,
  useRef,
  useState,
} from 'react';
import { Themes, TTheme, useTailwind } from '../utilities/tailwind';

/** Default legth of time to display tooltip after moving away from parent */
const DEFAULT_TIME = 1000;

/** Arrow box dimension */
const BOX_SIZE = 8;

/** Arrow box's hypotenuse length */
const FLOATING_OFFSET = Math.sqrt(2 * BOX_SIZE ** 2) / 2;

/**
 * Creates inline styles for arrows
 *
 * @param element - Arrow from middleware
 *
 * @param placement - Arrow placement
 *
 * @returns Arrow inline styles
 */
function createArrowStyles(
  element: MiddlewareData['arrow'],
  placement: Placement,
) {
  let styles = {
    bottom: '',
    left: '',
    right: '',
    top: '',
  };

  if (element) {
    const { x, y } = element;

    const side = placement.split('-')[0];

    /** Position to apply generated style */
    const xy = {
      bottom: 'top',
      left: 'right',
      right: 'left',
      top: 'bottom',
    }[side] as string;

    styles = {
      ...styles,
      left: xy === 'right' ? '' : `${x ?? BOX_SIZE}px`,
      top: xy === 'bottom' ? '' : `${y ?? BOX_SIZE}px`,
      [xy]: -BOX_SIZE / 2,
    };
  }

  return styles;
}

/** Tooltip component properties */
export interface ITooltipProps extends ComponentPropsWithoutRef<'div'> {
  /** Content to be displayed */
  content: ReactNode;

  /** Callback trigger when tooltop is shown or hidden  */
  onVisibilityChange?: (value: boolean) => void;

  /** Element to output tooltip relative to */
  parent: ReactNode;

  /** Where tooltip should be placed */
  position?: Side;

  /** Length of time to keep tooltip visible in milliseconds */
  time?: number;

  /** Component theme */
  theme?: TTheme;
}

/**
 * Display information when hovering over a component
 *
 * @param props - Component props
 *
 * @returns Tooltip component
 */
export function Tooltip({
  className,
  content,
  onVisibilityChange,
  parent,
  position = 'bottom',
  time = DEFAULT_TIME,
  theme = Themes.PRIMARY,
  ...rest
}: ITooltipProps) {
  const arrowRef = useRef(null);

  const { createStyles } = useTailwind(theme);

  const [isOpen, setIsOpen] = useState(false);

  const { context, middlewareData, placement, refs, strategy, x, y } =
    useFloating({
      middleware: [
        flip(),
        offset(FLOATING_OFFSET),
        shift(),
        arrow({
          element: arrowRef,
        }),
      ],
      onOpenChange: setIsOpen,
      open: isOpen,
      placement: position,
      whileElementsMounted: autoUpdate,
    });

  const hover = useHover(context, {
    delay: {
      close: time,
    },
    move: false,
  });

  const focus = useFocus(context);

  const dismiss = useDismiss(context);

  const role = useRole(context, {
    role: 'tooltip',
  });

  const { getFloatingProps, getReferenceProps } = useInteractions([
    hover,
    focus,
    dismiss,
    role,
  ]);

  const tooltipStyles = createStyles({
    classNames: [className],
    theme: {
      danger: ['bg-red-600 text-red-50'],
      primary: ['bg-blue-600 text-blue-50'],
      secondary: ['bg-gray-100 text-gray-600'],
      success: ['bg-green-600 text-green-50'],
      warning: ['bg-yellow-600 text-yellow-50'],
    },
  });

  const innerTooltipStyles = createStyles({
    classNames: [
      'px-1 text-sm font-display whitespace-nowrap',
      {
        'border-b-2': placement === 'top',
        'border-l-2': placement === 'right',
        'border-r-2': placement === 'left',
        'border-t-2': placement === 'bottom',
      },
    ],
    theme: {
      danger: ['border-red-700'],
      primary: ['border-blue-700'],
      secondary: ['border-gray-200'],
      success: ['border-green-700'],
      warning: ['border-yellow-700'],
    },
  });

  const tooltipInlineStyles = {
    left: x ?? 0,
    position: strategy,
    top: y ?? 0,
    width: 'max-content',
  };

  const arrowStyles = createStyles({
    classNames: ['absolute w-2 h-2 rotate-45 -z-10 pointer-events-none'],
    theme: {
      danger: ['bg-red-700'],
      primary: ['bg-blue-700'],
      secondary: ['bg-gray-200'],
      success: ['bg-green-700'],
      warning: ['bg-yellow-700'],
    },
  });

  const arrowInlineStyles = createArrowStyles(middlewareData.arrow, placement);

  useEffect(() => {
    if (onVisibilityChange) {
      onVisibilityChange(isOpen);
    }
  }, [isOpen, onVisibilityChange]);

  return (
    <>
      <div
        className="inline-block"
        data-testid="tooltip-parent-element"
        ref={refs.setReference}
        {...getReferenceProps()}
      >
        {parent}
      </div>

      {isOpen && (
        <div
          className={tooltipStyles}
          ref={refs.setFloating}
          style={tooltipInlineStyles}
          {...getFloatingProps()}
          {...rest}
        >
          <div className={innerTooltipStyles}>{content}</div>

          <div
            className={arrowStyles}
            ref={arrowRef}
            style={arrowInlineStyles}
          />
        </div>
      )}
    </>
  );
}
