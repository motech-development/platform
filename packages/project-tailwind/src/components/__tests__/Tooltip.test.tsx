import { Side } from '@floating-ui/react';
import { mockViewport } from 'jsdom-testing-mocks';
import { ReactNode } from 'react';
import { act } from 'react-dom/test-utils';
import { setup, themes } from '../../utilities/jest';
import { Tooltip } from '../Tooltip';

interface IWrapperProps {
  children: ReactNode;
}
function Wrapper({ children }: IWrapperProps) {
  const styles = {
    alignItems: 'center',
    display: 'flex',
    height: '100vh',
    justifyContent: 'center',
  };

  return <div style={styles}>{children}</div>;
}

const placements: Side[] = ['top', 'bottom', 'right', 'left'];

describe('Tooltip', () => {
  beforeAll(() => {
    mockViewport({
      height: '768px',
      width: '1024px',
    });
  });

  describe('timer', () => {
    beforeEach(() => {
      jest.useFakeTimers();
    });

    afterEach(() => {
      jest.useRealTimers();
    });

    it('should hide tooltip after 1000ms', async () => {
      const { asFragment, getByTestId, user } = setup(
        <Wrapper>
          <Tooltip
            parent={
              <button data-testid="button" type="button">
                Hover over me
              </button>
            }
          >
            <p>Hello, world</p>
          </Tooltip>
        </Wrapper>,
      );

      const firstRender = asFragment();

      await act(async () => {
        const parent = getByTestId('tooltip-parent-element');

        await user.hover(parent);

        await user.unhover(parent);
      });

      act(() => jest.advanceTimersByTime(900));

      expect(firstRender).toMatchDiffSnapshot(asFragment());

      act(() => jest.advanceTimersByTime(100));

      expect(firstRender).toEqual(asFragment());
    });

    it('should hide after set time', async () => {
      const { asFragment, getByTestId, user } = setup(
        <Wrapper>
          <Tooltip
            parent={
              <button data-testid="button" type="button">
                Hover over me
              </button>
            }
            time={400}
          >
            <p>Hello, world</p>
          </Tooltip>
        </Wrapper>,
      );

      const firstRender = asFragment();

      await act(async () => {
        const parent = getByTestId('tooltip-parent-element');

        await user.hover(parent);

        await user.unhover(parent);
      });

      act(() => jest.advanceTimersByTime(300));

      expect(firstRender).toMatchDiffSnapshot(asFragment());

      act(() => jest.advanceTimersByTime(100));

      expect(firstRender).toEqual(asFragment());
    });
  });

  describe.each(placements)('when position is "%s"', (position) => {
    it('should correctly render component when parent is not hovered over', () => {
      const { asFragment } = setup(
        <Wrapper>
          <Tooltip
            parent={
              <button data-testid="button" type="button">
                Hover over me
              </button>
            }
            position={position}
          >
            <p>Hello, world</p>
          </Tooltip>
        </Wrapper>,
      );

      expect(asFragment()).toMatchSnapshot();
    });

    it('should correctly render component when parent is hovered over', async () => {
      const { asFragment, getByTestId, user } = setup(
        <Wrapper>
          <Tooltip
            parent={
              <button data-testid="button" type="button">
                Hover over me
              </button>
            }
            position={position}
          >
            <p>Hello, world</p>
          </Tooltip>
        </Wrapper>,
      );

      const firstRender = asFragment();

      await act(async () => {
        const parent = getByTestId('tooltip-parent-element');

        await user.hover(parent);
      });

      expect(firstRender).toMatchDiffSnapshot(asFragment());
    });
  });

  describe.each(themes)('when theme is "$theme"', ({ theme }) => {
    it('should correctly render component when parent is not hovered over', () => {
      const { asFragment } = setup(
        <Wrapper>
          <Tooltip
            parent={
              <button data-testid="button" type="button">
                Hover over me
              </button>
            }
            theme={theme}
          >
            <p>Hello, world</p>
          </Tooltip>
        </Wrapper>,
      );

      expect(asFragment()).toMatchSnapshot();
    });

    it('should correctly render component when parent is hovered over', async () => {
      const { asFragment, getByTestId, user } = setup(
        <Wrapper>
          <Tooltip
            parent={
              <button data-testid="button" type="button">
                Hover over me
              </button>
            }
            theme={theme}
          >
            <p>Hello, world</p>
          </Tooltip>
        </Wrapper>,
      );

      const firstRender = asFragment();

      await act(async () => {
        const parent = getByTestId('tooltip-parent-element');

        await user.hover(parent);
      });

      expect(firstRender).toMatchDiffSnapshot(asFragment());
    });
  });
});
