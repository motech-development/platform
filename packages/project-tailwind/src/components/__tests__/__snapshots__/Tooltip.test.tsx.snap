// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip timer should hide after set time 1`] = `
"Snapshot Diff:
- First value
+ Second value

  <DocumentFragment>
    <div
      style="align-items: center; display: flex; height: 100vh; justify-content: center;"
    >
      <div
+       aria-describedby="floating-ui-2"
        class="inline-block"
        data-testid="tooltip-parent-element"
      >
        <button
          data-testid="button"
          type="button"
        >
          Hover over me
        </button>
+     </div>
+     <div
+       class="bg-blue-600 text-blue-50"
+       id="floating-ui-2"
+       role="tooltip"
+       style="left: 0px; position: absolute; top: 5.656854249492381px;"
+       tabindex="-1"
+     >
+       <div
+         class="px-1 text-sm font-display whitespace-nowrap border-t-2 border-blue-700"
+       >
+         <p>
+           Hello, world
+         </p>
+       </div>
+       <div
+         class="absolute w-2 h-2 rotate-45 -z-10 pointer-events-none bg-blue-700"
+         style="left: 0px; top: -4px;"
+       />
      </div>
    </div>
  </DocumentFragment>"
`;

exports[`Tooltip timer should hide tooltip after 1000ms 1`] = `
"Snapshot Diff:
- First value
+ Second value

  <DocumentFragment>
    <div
      style="align-items: center; display: flex; height: 100vh; justify-content: center;"
    >
      <div
+       aria-describedby="floating-ui-0"
        class="inline-block"
        data-testid="tooltip-parent-element"
      >
        <button
          data-testid="button"
          type="button"
        >
          Hover over me
        </button>
+     </div>
+     <div
+       class="bg-blue-600 text-blue-50"
+       id="floating-ui-0"
+       role="tooltip"
+       style="left: 0px; position: absolute; top: 5.656854249492381px;"
+       tabindex="-1"
+     >
+       <div
+         class="px-1 text-sm font-display whitespace-nowrap border-t-2 border-blue-700"
+       >
+         <p>
+           Hello, world
+         </p>
+       </div>
+       <div
+         class="absolute w-2 h-2 rotate-45 -z-10 pointer-events-none bg-blue-700"
+         style="left: 0px; top: -4px;"
+       />
      </div>
    </div>
  </DocumentFragment>"
`;

exports[`Tooltip when placement is "bottom" should correctly render component when parent is hovered over 1`] = `
"Snapshot Diff:
- First value
+ Second value

  <DocumentFragment>
    <div
      style="align-items: center; display: flex; height: 100vh; justify-content: center;"
    >
      <div
+       aria-describedby="floating-ui-10"
        class="inline-block"
        data-testid="tooltip-parent-element"
      >
        <button
          data-testid="button"
          type="button"
        >
          Hover over me
        </button>
+     </div>
+     <div
+       class="bg-blue-600 text-blue-50"
+       id="floating-ui-10"
+       role="tooltip"
+       style="left: 0px; position: absolute; top: 5.656854249492381px;"
+       tabindex="-1"
+     >
+       <div
+         class="px-1 text-sm font-display whitespace-nowrap border-t-2 border-blue-700"
+       >
+         <p>
+           Hello, world
+         </p>
+       </div>
+       <div
+         class="absolute w-2 h-2 rotate-45 -z-10 pointer-events-none bg-blue-700"
+         style="left: 0px; top: -4px;"
+       />
      </div>
    </div>
  </DocumentFragment>"
`;

exports[`Tooltip when placement is "bottom" should correctly render component when parent is not hovered over 1`] = `
<DocumentFragment>
  <div
    style="align-items: center; display: flex; height: 100vh; justify-content: center;"
  >
    <div
      class="inline-block"
      data-testid="tooltip-parent-element"
    >
      <button
        data-testid="button"
        type="button"
      >
        Hover over me
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Tooltip when placement is "left" should correctly render component when parent is hovered over 1`] = `
"Snapshot Diff:
- First value
+ Second value

  <DocumentFragment>
    <div
      style="align-items: center; display: flex; height: 100vh; justify-content: center;"
    >
      <div
+       aria-describedby="floating-ui-18"
        class="inline-block"
        data-testid="tooltip-parent-element"
      >
        <button
          data-testid="button"
          type="button"
        >
          Hover over me
        </button>
+     </div>
+     <div
+       class="bg-blue-600 text-blue-50"
+       id="floating-ui-18"
+       role="tooltip"
+       style="left: -5.656854249492381px; position: absolute; top: 0px;"
+       tabindex="-1"
+     >
+       <div
+         class="px-1 text-sm font-display whitespace-nowrap border-r-2 border-blue-700"
+       >
+         <p>
+           Hello, world
+         </p>
+       </div>
+       <div
+         class="absolute w-2 h-2 rotate-45 -z-10 pointer-events-none bg-blue-700"
+         style="right: -4px; top: 0px;"
+       />
      </div>
    </div>
  </DocumentFragment>"
`;

exports[`Tooltip when placement is "left" should correctly render component when parent is not hovered over 1`] = `
<DocumentFragment>
  <div
    style="align-items: center; display: flex; height: 100vh; justify-content: center;"
  >
    <div
      class="inline-block"
      data-testid="tooltip-parent-element"
    >
      <button
        data-testid="button"
        type="button"
      >
        Hover over me
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Tooltip when placement is "right" should correctly render component when parent is hovered over 1`] = `
"Snapshot Diff:
- First value
+ Second value

  <DocumentFragment>
    <div
      style="align-items: center; display: flex; height: 100vh; justify-content: center;"
    >
      <div
+       aria-describedby="floating-ui-14"
        class="inline-block"
        data-testid="tooltip-parent-element"
      >
        <button
          data-testid="button"
          type="button"
        >
          Hover over me
        </button>
+     </div>
+     <div
+       class="bg-blue-600 text-blue-50"
+       id="floating-ui-14"
+       role="tooltip"
+       style="left: 5.656854249492381px; position: absolute; top: 0px;"
+       tabindex="-1"
+     >
+       <div
+         class="px-1 text-sm font-display whitespace-nowrap border-l-2 border-blue-700"
+       >
+         <p>
+           Hello, world
+         </p>
+       </div>
+       <div
+         class="absolute w-2 h-2 rotate-45 -z-10 pointer-events-none bg-blue-700"
+         style="left: -4px; top: 0px;"
+       />
      </div>
    </div>
  </DocumentFragment>"
`;

exports[`Tooltip when placement is "right" should correctly render component when parent is not hovered over 1`] = `
<DocumentFragment>
  <div
    style="align-items: center; display: flex; height: 100vh; justify-content: center;"
  >
    <div
      class="inline-block"
      data-testid="tooltip-parent-element"
    >
      <button
        data-testid="button"
        type="button"
      >
        Hover over me
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Tooltip when placement is "top" should correctly render component when parent is hovered over 1`] = `
"Snapshot Diff:
- First value
+ Second value

  <DocumentFragment>
    <div
      style="align-items: center; display: flex; height: 100vh; justify-content: center;"
    >
      <div
+       aria-describedby="floating-ui-6"
        class="inline-block"
        data-testid="tooltip-parent-element"
      >
        <button
          data-testid="button"
          type="button"
        >
          Hover over me
        </button>
+     </div>
+     <div
+       class="bg-blue-600 text-blue-50"
+       id="floating-ui-6"
+       role="tooltip"
+       style="left: 0px; position: absolute; top: -5.656854249492381px;"
+       tabindex="-1"
+     >
+       <div
+         class="px-1 text-sm font-display whitespace-nowrap border-b-2 border-blue-700"
+       >
+         <p>
+           Hello, world
+         </p>
+       </div>
+       <div
+         class="absolute w-2 h-2 rotate-45 -z-10 pointer-events-none bg-blue-700"
+         style="bottom: -4px; left: 0px;"
+       />
      </div>
    </div>
  </DocumentFragment>"
`;

exports[`Tooltip when placement is "top" should correctly render component when parent is not hovered over 1`] = `
<DocumentFragment>
  <div
    style="align-items: center; display: flex; height: 100vh; justify-content: center;"
  >
    <div
      class="inline-block"
      data-testid="tooltip-parent-element"
    >
      <button
        data-testid="button"
        type="button"
      >
        Hover over me
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Tooltip when theme is "danger" should correctly render component when parent is hovered over 1`] = `
"Snapshot Diff:
- First value
+ Second value

  <DocumentFragment>
    <div
      style="align-items: center; display: flex; height: 100vh; justify-content: center;"
    >
      <div
+       aria-describedby="floating-ui-22"
        class="inline-block"
        data-testid="tooltip-parent-element"
      >
        <button
          data-testid="button"
          type="button"
        >
          Hover over me
        </button>
+     </div>
+     <div
+       class="bg-red-600 text-red-50"
+       id="floating-ui-22"
+       role="tooltip"
+       style="left: 0px; position: absolute; top: 5.656854249492381px;"
+       tabindex="-1"
+     >
+       <div
+         class="px-1 text-sm font-display whitespace-nowrap border-t-2 border-red-700"
+       >
+         <p>
+           Hello, world
+         </p>
+       </div>
+       <div
+         class="absolute w-2 h-2 rotate-45 -z-10 pointer-events-none bg-red-700"
+         style="left: 0px; top: -4px;"
+       />
      </div>
    </div>
  </DocumentFragment>"
`;

exports[`Tooltip when theme is "danger" should correctly render component when parent is not hovered over 1`] = `
<DocumentFragment>
  <div
    style="align-items: center; display: flex; height: 100vh; justify-content: center;"
  >
    <div
      class="inline-block"
      data-testid="tooltip-parent-element"
    >
      <button
        data-testid="button"
        type="button"
      >
        Hover over me
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Tooltip when theme is "primary" should correctly render component when parent is hovered over 1`] = `
"Snapshot Diff:
- First value
+ Second value

  <DocumentFragment>
    <div
      style="align-items: center; display: flex; height: 100vh; justify-content: center;"
    >
      <div
+       aria-describedby="floating-ui-26"
        class="inline-block"
        data-testid="tooltip-parent-element"
      >
        <button
          data-testid="button"
          type="button"
        >
          Hover over me
        </button>
+     </div>
+     <div
+       class="bg-blue-600 text-blue-50"
+       id="floating-ui-26"
+       role="tooltip"
+       style="left: 0px; position: absolute; top: 5.656854249492381px;"
+       tabindex="-1"
+     >
+       <div
+         class="px-1 text-sm font-display whitespace-nowrap border-t-2 border-blue-700"
+       >
+         <p>
+           Hello, world
+         </p>
+       </div>
+       <div
+         class="absolute w-2 h-2 rotate-45 -z-10 pointer-events-none bg-blue-700"
+         style="left: 0px; top: -4px;"
+       />
      </div>
    </div>
  </DocumentFragment>"
`;

exports[`Tooltip when theme is "primary" should correctly render component when parent is not hovered over 1`] = `
<DocumentFragment>
  <div
    style="align-items: center; display: flex; height: 100vh; justify-content: center;"
  >
    <div
      class="inline-block"
      data-testid="tooltip-parent-element"
    >
      <button
        data-testid="button"
        type="button"
      >
        Hover over me
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Tooltip when theme is "secondary" should correctly render component when parent is hovered over 1`] = `
"Snapshot Diff:
- First value
+ Second value

  <DocumentFragment>
    <div
      style="align-items: center; display: flex; height: 100vh; justify-content: center;"
    >
      <div
+       aria-describedby="floating-ui-30"
        class="inline-block"
        data-testid="tooltip-parent-element"
      >
        <button
          data-testid="button"
          type="button"
        >
          Hover over me
        </button>
+     </div>
+     <div
+       class="bg-gray-100 text-gray-600"
+       id="floating-ui-30"
+       role="tooltip"
+       style="left: 0px; position: absolute; top: 5.656854249492381px;"
+       tabindex="-1"
+     >
+       <div
+         class="px-1 text-sm font-display whitespace-nowrap border-t-2 border-gray-200"
+       >
+         <p>
+           Hello, world
+         </p>
+       </div>
+       <div
+         class="absolute w-2 h-2 rotate-45 -z-10 pointer-events-none bg-gray-200"
+         style="left: 0px; top: -4px;"
+       />
      </div>
    </div>
  </DocumentFragment>"
`;

exports[`Tooltip when theme is "secondary" should correctly render component when parent is not hovered over 1`] = `
<DocumentFragment>
  <div
    style="align-items: center; display: flex; height: 100vh; justify-content: center;"
  >
    <div
      class="inline-block"
      data-testid="tooltip-parent-element"
    >
      <button
        data-testid="button"
        type="button"
      >
        Hover over me
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Tooltip when theme is "success" should correctly render component when parent is hovered over 1`] = `
"Snapshot Diff:
- First value
+ Second value

  <DocumentFragment>
    <div
      style="align-items: center; display: flex; height: 100vh; justify-content: center;"
    >
      <div
+       aria-describedby="floating-ui-34"
        class="inline-block"
        data-testid="tooltip-parent-element"
      >
        <button
          data-testid="button"
          type="button"
        >
          Hover over me
        </button>
+     </div>
+     <div
+       class="bg-green-600 text-green-50"
+       id="floating-ui-34"
+       role="tooltip"
+       style="left: 0px; position: absolute; top: 5.656854249492381px;"
+       tabindex="-1"
+     >
+       <div
+         class="px-1 text-sm font-display whitespace-nowrap border-t-2 border-green-700"
+       >
+         <p>
+           Hello, world
+         </p>
+       </div>
+       <div
+         class="absolute w-2 h-2 rotate-45 -z-10 pointer-events-none bg-green-700"
+         style="left: 0px; top: -4px;"
+       />
      </div>
    </div>
  </DocumentFragment>"
`;

exports[`Tooltip when theme is "success" should correctly render component when parent is not hovered over 1`] = `
<DocumentFragment>
  <div
    style="align-items: center; display: flex; height: 100vh; justify-content: center;"
  >
    <div
      class="inline-block"
      data-testid="tooltip-parent-element"
    >
      <button
        data-testid="button"
        type="button"
      >
        Hover over me
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Tooltip when theme is "warning" should correctly render component when parent is hovered over 1`] = `
"Snapshot Diff:
- First value
+ Second value

  <DocumentFragment>
    <div
      style="align-items: center; display: flex; height: 100vh; justify-content: center;"
    >
      <div
+       aria-describedby="floating-ui-38"
        class="inline-block"
        data-testid="tooltip-parent-element"
      >
        <button
          data-testid="button"
          type="button"
        >
          Hover over me
        </button>
+     </div>
+     <div
+       class="bg-yellow-600 text-yellow-50"
+       id="floating-ui-38"
+       role="tooltip"
+       style="left: 0px; position: absolute; top: 5.656854249492381px;"
+       tabindex="-1"
+     >
+       <div
+         class="px-1 text-sm font-display whitespace-nowrap border-t-2 border-yellow-700"
+       >
+         <p>
+           Hello, world
+         </p>
+       </div>
+       <div
+         class="absolute w-2 h-2 rotate-45 -z-10 pointer-events-none bg-yellow-700"
+         style="left: 0px; top: -4px;"
+       />
      </div>
    </div>
  </DocumentFragment>"
`;

exports[`Tooltip when theme is "warning" should correctly render component when parent is not hovered over 1`] = `
<DocumentFragment>
  <div
    style="align-items: center; display: flex; height: 100vh; justify-content: center;"
  >
    <div
      class="inline-block"
      data-testid="tooltip-parent-element"
    >
      <button
        data-testid="button"
        type="button"
      >
        Hover over me
      </button>
    </div>
  </div>
</DocumentFragment>
`;
